src/execution/__tests__/subscribe-test.ts

```diff
@@ -165,13 +164,28 @@ async function subscribeWithBadFn(
     }),
   });
   const document = parse('subscription { foo }');
-  const result = await subscribe({ schema, document });
 
-  assert(!isAsyncIterable(result));
-  expectJSON(await createSourceEventStream(schema, document)).toDeepEqual(
-    result,
-  );
-  return result;
+  const subscribeResult = subscribe({ schema, document });
+  const streamResult = createSourceEventStream(schema, document);
+
+  if (isPromise(subscribeResult)) {
+    assert(isPromise(streamResult));
+    return Promise.all([subscribeResult, streamResult]).then((resolved) =>
+      expectEquivalentStreamErrors(resolved[0], resolved[1]),
+    );
+  }
+
+  assert(!isPromise(streamResult));
+  return expectEquivalentStreamErrors(subscribeResult, streamResult);
+}
+
+function expectEquivalentStreamErrors(
+  subscribeResult: ExecutionResult | AsyncGenerator<ExecutionResult>,
+  createSourceEventStreamResult: ExecutionResult | AsyncIterable<unknown>,
+): ExecutionResult {
+  assert(!isAsyncIterable(subscribeResult));
+  expectJSON(createSourceEventStreamResult).toDeepEqual(subscribeResult);
+  return subscribeResult;
```

Comment:
I understand why this code is needed to preserve old tests intact.
In a long run, it's too much complexity for tests.
I'm ok merging it for now but please try to simplify it in a separate PR.

Or if you think this code will change with upcoming refactoring please feel free to leave a 'TODO' comment and finish it after you finish with refactoring.

Reply:
Simplified tests in [f88b523](https://github.com/graphql/graphql-js/pull/3620/commits/f88b523c175b1c1280dc4659be12426f8f9ea07f)

Reply:
And added TODOs